src/compile/data/optimize.ts

```diff
@@ -127,45 +140,71 @@ export function mergeParse(node: DataFlowNode) {
       }
     }
     if (keys(commonParse).length !== 0) {
-      const mergedParseNode = new ParseNode(node, commonParse);
+      flag = true;
+      const mergedParseNode = new ParseNode(parent, commonParse);
       for (const parseNode of parseChildren) {
         for (const key of keys(commonParse)) {
           delete parseNode.parse[key];
         }
-        node.removeChild(parseNode);
+        parent.removeChild(parseNode);
         parseNode.parent = mergedParseNode;
         if (keys(parseNode.parse).length === 0) {
           parseNode.remove();
         }
       }
     }
   }
-  node.children.forEach(mergeParse);
+  return {continueFlag: true, mutatedFlag: flag};
 }
 
-/**
- * Optimizes the dataflow of the passed in data component.
- */
-export function optimizeDataflow(dataComponent: DataComponent) {
+function optimizationDataflowHelper(dataComponent: DataComponent) {
   let roots: SourceNode[] = vals(dataComponent.sources);
-
-  roots.forEach(removeUnnecessaryNodes);
-
+  let mutatedFlag = false;
+  // mutatedFlag should always be on the right side otherwise short circuit logic might cause the mutating method to not execute
+  mutatedFlag = roots.map(removeUnnecessaryNodes).some(x => x === true) || mutatedFlag;
   // remove source nodes that don't have any children because they also don't have output nodes
   roots = roots.filter(r => r.numChildren() > 0);
-  getLeaves(roots).forEach(optimizers.iterateFromLeaves(optimizers.removeUnusedSubtrees));
+
+  mutatedFlag =
+    getLeaves(roots)
+      .map(optimizers.iterateFromLeaves(optimizers.removeUnusedSubtrees))
+      .some(x => x === true) || mutatedFlag;
+
   roots = roots.filter(r => r.numChildren() > 0);
 
-  getLeaves(roots).forEach(optimizers.iterateFromLeaves(optimizers.moveParseUp));
-  getLeaves(roots).forEach(optimizers.removeDuplicateTimeUnits);
+  mutatedFlag =
+    getLeaves(roots)
+      .map(optimizers.iterateFromLeaves(optimizers.moveParseUp))
+      .some(x => x === true) || mutatedFlag;
+
+  mutatedFlag =
+    getLeaves(roots)
+      .map(optimizers.iterateFromLeaves(mergeParse))
+      .some(x => x === true) || mutatedFlag;
+
+  mutatedFlag =
+    getLeaves(roots)
+      .map(optimizers.removeDuplicateTimeUnits)
+      .some(x => x === true) || mutatedFlag;
 
-  roots.forEach(moveFacetDown);
-  roots.forEach(mergeParse);
-  roots.forEach(optimizers.mergeIdenticalTransforms);
+  mutatedFlag = roots.map(moveFacetDown).some(x => x === true) || mutatedFlag;
+  mutatedFlag = roots.map(optimizers.mergeIdenticalTransforms).some(x => x === true) || mutatedFlag;
 
   keys(dataComponent.sources).forEach(s => {
     if (dataComponent.sources[s].numChildren() === 0) {
       delete dataComponent.sources[s];
     }
   });
+  return mutatedFlag;
+}
+
+/**
+ * Optimizes the dataflow of the passed in data component.
+ */
+export function optimizeDataflow(data: DataComponent) {
+  for (let i = 0; i < 5; i++) {
```

Comment:
Having a constant like `MAX_OPTIMIZATION_RUNS` would be more readable than having the hard-coded 5.

Reply:
Done.